[16.0][ADD] account_edi_ubl_cii_purchase_match

sbejaoui opened the pull request:
Extend UBL vendor bill import to automatically match and link bill lines to purchase order lines using the OrderReference and product label

sbejaoui committed:
[ADD] account_edi_ubl_cii_purchase_match

Extend UBL vendor bill import to automatically match and link bill lines to purchase order lines using the OrderReference and product label

[IMP] account_edi_ubl_cii_purchase_match: show price_subtotal in purchase line form view

[FIX] account_edi_ubl_cii_purchase_match: allow to change po line

[FIX] account_edi_ubl_cii_purchase_match: don't update supplier info product name

[FIX] account_edi_ubl_cii_purchase_match: match po only if order_ref is set

sbejaoui committed:
[ADD] account_edi_ubl_cii_purchase_match

Extend UBL vendor bill import to automatically match and link bill lines to purchase order lines using the OrderReference and product label

[IMP] account_edi_ubl_cii_purchase_match: show price_subtotal in purchase line form view

[FIX] account_edi_ubl_cii_purchase_match: allow to change po line

[FIX] account_edi_ubl_cii_purchase_match: don't update supplier info product name

[FIX] account_edi_ubl_cii_purchase_match: match po only if order_ref is set

sbejaoui commented:
cc/ @phschmidt , @jbaudoux , @rousseldenis

phschmidt reviewed (approved):
Functional review

jbaudoux commented on account_edi_ubl_cii_purchase_match/models/account_edi_xml_ubl_20.py:
```diff
@@ -0,0 +1,111 @@
+            [
+                "|",
+                ("name", "=", order_ref),
+                ("partner_ref", "=", order_ref),
```
`partner_ref` should compare to `OrderReference/SalesOrderID` and not  `OrderReference/ID`

sbejaoui replied:
I will use what base module stored in invoice_origin

jbaudoux commented on account_edi_ubl_cii_purchase_match/models/account_edi_xml_ubl_20.py:
```diff
@@ -0,0 +1,111 @@
+        res = super()._import_fill_invoice_form(journal, tree, invoice, qty_factor)
+        if journal.type != "purchase":
+            return res
+        invoice_origin_node = tree.find("./{*}OrderReference/{*}ID")
```
Maybe you should only search for the parent here as ID is mandatory when the parent is there and then parse ID and SalesOrderID

jbaudoux commented on account_edi_ubl_cii_purchase_match/models/account_edi_xml_ubl_20.py:
```diff
@@ -0,0 +1,111 @@
+        invoice_origin_node = tree.find("./{*}OrderReference/{*}ID")
+        if invoice_origin_node is None:
+            return res
+        order_ref = invoice_origin_node.text
```
then handover this to the called method below

jbaudoux commented on account_edi_ubl_cii_purchase_match/models/account_edi_xml_ubl_20.py:
```diff
@@ -0,0 +1,111 @@
+        if invoice_line.purchase_line_id:
+            return False
+        product = self._get_matching_product(invoice_line)
+        purchase_line = purchase_lines.filtered(lambda line: line.product_id == product)
```
What if no product is found and there are PO lines without product?
```suggestion
        purchase_line = purchase_lines.filtered(lambda line: line.product_id and line.product_id == product)
```

jbaudoux commented on account_edi_ubl_cii_purchase_match/models/account_edi_xml_ubl_20.py:
```diff
@@ -0,0 +1,111 @@
+        supplier_product_code = self._find_value(
+            "./cac:Item/cac:SellersItemIdentification/cbc:ID", tree
+        )
+        if invoice_line.product_id or not supplier_product_code:
```
There is I think a bug in the call to super that retrieves the product. It's searching for a product with default_code based on the SellersItemIdentification while it should be BuyersItemIdentification
https://github.com/odoo/odoo/blob/f07949e4ccdb5d7a17a6678b008fa6785ec3fb12/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py#L727
I would force a new call to the method `_retrieve_product` with the right argument

jbaudoux commented on account_edi_ubl_cii_purchase_match/models/account_edi_xml_ubl_20.py:
```diff
@@ -0,0 +1,111 @@
+        invoice_line.purchase_line_id = purchase_line
+        return True
+
+    def _get_matching_product(self, invoice_line):
```
What's the purpose of this? The product matching should already be resolved by `_retrieve_product` and what you improved in `_import_fill_invoice_line_form`?

What could be usefull is to match a PO line name based on the invoice line name, in case the seller or buyer does not manage the line with a product

jbaudoux commented:
@sbejaoui Can you attend comments?

sbejaoui commented:
pipeline is red due to another module #1313

sbejaoui committed:
[IMP] account_edi_ubl_cii_purchase_match: support multiple purchase orders

Extract multiple purchase order references from invoice_origin using
a prefix-based pattern and match them with existing purchase orders

sbejaoui committed:
[IMP] account_edi_ubl_cii_purchase_match: support multiple purchase orders

Extract multiple purchase order references from invoice_origin using
a prefix-based pattern and match them with existing purchase orders

lmignon reviewed (approved):
code review + functional tests

sbejaoui committed:
[IMP] account_edi_ubl_cii_purchase_match: detect invoiced qty above received qty

sbejaoui commented:
/ocabot merge nobump

sbejaoui commented:
/ocabot merge nobump

sbejaoui commented:
/ocabot merge nobump

sbejaoui commented:
/ocabot merge nobump

sbejaoui commented:
/ocabot merge nobump
